Mage.Sets/src/mage/cards/a/AnrakyrTheTraveller.java

```diff
@@ -0,0 +1,198 @@
+    }
+
+    
+
+    @Override
+    public boolean apply(Game game, Ability source) {
+        Player player = game.getPlayer(source.getControllerId());
+        if (player == null) {
+            return false;
+        }
+        
+        Set<Card> cards = player.getHand().getCards(filter, game);
+        cards.addAll(player.getGraveyard().getCards(filter, game));
+        
+        Map<UUID, List<Card>> cardMap = new HashMap<>();
+        for (Card card : cards) {
+            List<Card> castableComponents = getCastableComponents(card, filter, source, player, game, false);
+            if (!castableComponents.isEmpty()) {
+                cardMap.put(card.getId(), castableComponents);
+            }
+        }      
+        Card cardToCast;
+        switch (cardMap.size()) {
+            case 0:
+                return false;
+            default:
+                Cards castableCards = new CardsImpl(cardMap.keySet());
+                TargetCard target = new TargetCard(0, 1, Zone.ALL, filter);
+                target.withNotTarget(true);
+                player.choose(Outcome.Benefit, castableCards, target, source, game);
+                cardToCast = castableCards.get(target.getFirstTarget(), game);
+        }  
+
+        if (cardToCast == null) {
+            return false;
+        }
+
+
+        List<Card> partsToCast = cardMap.get(cardToCast.getId());
+        String partsInfo = partsToCast
+                .stream()
+                .map(MageObject::getIdName)
+                .collect(Collectors.joining(" or "));
+        if (cardToCast == null
+                || partsToCast.size() < 1
+                || !player.chooseUse(
+                Outcome.PlayForFree, "Cast spell by paying life equal to its mana value rather than paying its mana cost (" + partsInfo + ")?", source, game
+        )) {
+            return false;
+        }
+        partsToCast.forEach(card -> game.getState().setValue("PlayFromNotOwnHandZone" + card.getId(), Boolean.TRUE));
+        
+        // pay life
+        PayLifeCost lifeCost = new PayLifeCost(cardToCast.getSpellAbility().getManaCosts().manaValue());
+        Costs<Cost> newCosts = new CostsImpl<>();
+        newCosts.add(lifeCost);
+        newCosts.addAll(cardToCast.getSpellAbility().getCosts());
+        player.setCastSourceIdWithAlternateMana(cardToCast.getId(), null, newCosts);
+
+        ActivatedAbility chosenAbility;
+        chosenAbility = player.chooseAbilityForCast(cardToCast, game, true);
+        boolean result = false;
+        if (chosenAbility instanceof SpellAbility) {
+            result = player.cast(
+                    (SpellAbility) chosenAbility,
+                    game, true, new ApprovingObject(source, game)
+            );
+        }
+        partsToCast.forEach(card -> game.getState().setValue("PlayFromNotOwnHandZone" + card.getId(), null));
+        if (player.isComputer() && !result) {
+            cards.remove(cardToCast);
+        }
```

Comment:
What exactly is that for?

Reply:
We copied this from `mage.util.CardUtil::castSpellWithAttributesForFree`. Is it not necessary in this case?

Reply:
@theelk801 do you have any recollection of why this exists? seems pointless to me

https://github.com/magefree/mage/blob/845b9da215873cf025ca1cc5bfb513e9cd6ae7e8/Mage/src/main/java/mage/util/CardUtil.java#L1391-L1393

Reply:
I believe it was added per @JayDi85's request to prevent the computer from accidentally causing a loop

Reply:
Alright then, thanks

Reply:
(wish that were a comment in the code itself)

Reply:
Computer can select wrong choice (e.g. protected target), but it can’t understand it and will choose again with same choice. Its will cause AI loop and game freeze. So all “while cycles” (e.g. infinite choose) must contain infinite loop protection like that (you give AI only one attempt per choice).